### PR TITLE
KAZOO-5153: make templates great again

### PR DESCRIPTION
**Pull request by icehess**
* Style Templates
* Fix some missing information on templates
* Bring back some templates to life
* Change to transaction: break it into two different templates(one for success, one for payment problems)
* some macro names in templates are changed (port_*, service_added for example)

### CODE REVIEW

File: core/braintree/src/braintree_transaction.erl [hunk 1]
```diff
@@ -494,3 +494,29 @@ json_to_record(JObj) ->
                    ,is_automatic = kz_json:is_true(<<"is_automatic">>, JObj)
                    ,is_recurring = kz_json:is_true(<<"is_recurring">>, JObj)
       }.
+
+-spec record_to_notification_props(bt_transaction()) -> kz_proplist().
```
Comment (jamesaimonetti) [inline core/braintree/src/braintree_transaction.erl]: Unexported / unused function

File: core/braintree/src/braintree_transaction.erl [hunk 2]
```diff
@@ -494,3 +494,29 @@ json_to_record(JObj) ->
...
+      ,{<<"ID">>, kz_json:get_value(<<"id">>, Transaction)}
+      ,{<<"Add-Ons">>, kz_json:get_value(<<"add_ons">>, Transaction)}
+      ,{<<"Discounts">>, kz_json:get_value(<<"discounts">>, Transaction)}
+      ,{<<"Billing-Address">>, kz_json:get_value(<<"billing_address">>, Transaction)}
+      ,{<<"Card-Last-Four">>, kz_json:get_value([<<"card">>, <<"last_four">>], Transaction)}
+      ,{<<"Tax-Amount">>, kz_json:get_value(<<"tax_amount">>, Transaction)}
+      ,{<<"Timestamp">>, kz_time:current_tstamp()}
+      ,{<<"Purchase-Order">>, purchase_order_reason(Transaction)}
+      ,{<<"Currency-Code">>, kz_json:get_value(<<"currency_code">>, Transaction)}
+      ].
```
Comment (jamesaimonetti) [inline core/braintree/src/braintree_transaction.erl]: Missing the closing paren for `props:filter_empty/2`

File: core/braintree/src/braintree_transaction.erl [hunk 3]
```diff
@@ -494,3 +494,29 @@ json_to_record(JObj) ->
                    ,is_automatic = kz_json:is_true(<<"is_automatic">>, JObj)
                    ,is_recurring = kz_json:is_true(<<"is_recurring">>, JObj)
       }.
+
+-spec record_to_notification_props(bt_transaction()) -> kz_proplist().
+record_to_notification_props(#bt_transaction{}=BraintreeTransaction) ->
+    Transaction = record_to_json(BraintreeTransaction),
+    RespCode = kz_json:get_value(<<"processor_response_code">>, Transaction, ?CODE_UNKNOWN),
```
Comment (jamesaimonetti) [inline core/braintree/src/braintree_transaction.erl]: Macro doesn't exist unless you include `kazoo_transactions.hrl`

File: applications/teletype/src/templates/teletype_service_added.erl [hunk 4]
```diff
@@ -98,32 +107,18 @@ reseller_info_data(DataJObj) ->
...
             {'ok', AccountJObj} = kz_account:fetch(AccountId),
-            props:filter_undefined(
-              [{<<"name">>, kz_account:name(AccountJObj)}
-              ,{<<"id">>, AccountId}
-              ,{<<"realm">>, kz_account:realm(AccountJObj)}
-              ,{<<"language">>, kz_account:language(AccountJObj)}
-              ,{<<"timezone">>, kz_account:timezone(AccountJObj)}
-              ]
-             )
+            teletype_util:find_account_params(DataJObj, ResellerId)
```
Comment (jamesaimonetti) [inline applications/teletype/src/templates/teletype_service_added.erl]: ResellerId is undefined

**Comment by jamesaimonetti** [comment]
@icehess this does not seem like a PR ready for review (lots of compilation errors). Can you close it and reopen when you're ready to have it reviewed? If you want to see CI reports, you can switch to the branch, check the commits list, and there should be the yellow/green/red icon next to the latest commit that will take you to the tests being run.

**Comment by jamhed** [comment]
I like the pr header.

File: core/braintree/src/braintree_transaction.erl (hunk 1 above)
Comment (icehess) [inline core/braintree/src/braintree_transaction.erl]: I fixed that last night but didn't commit, it's now pushed

File: core/braintree/src/braintree_transaction.erl (hunk 3 above)
Comment (icehess) [inline core/braintree/src/braintree_transaction.erl]: same as above

File: applications/teletype/src/templates/teletype_service_added.erl (hunk 4 above)
Comment (icehess) [inline applications/teletype/src/templates/teletype_service_added.erl]: fixed

**Comment by fenollp** [comment]
While the PR does not mention it, this is still ongoing and maybe not ready for review.
But there's no need to close & reopen PR. There's the chance that someone comes around and deletes the branch or something too.

**Comment by icehess** [comment]
Except System alert templates, it's ready. You can merge this PR if test passed, no need for waiting to system_alert since this is only would be sent to system admin and there are lots of data dump in it which is really difficult to style it properly.

I'll return it to at a later time.

**Comment by fenollp** [comment]
@icehess some Dialyzer complaints ;)
You should rebase before fixing them though, should remove some of them.

**Comment by icehess** [comment]
Dialyzer warnings are not related to changes in this PR, they're also exists on the master too. They're related to `bt_transaction{}` record construction which in most places are violating the declared types.

**Comment by fenollp** [comment]
I we merge PRs with failing Dialyzer tests this will impact every PRs. We should fix this

**Comment by k-anderson** [comment]
@jamesaimonetti hesaam says he addressed your comments, if so can you close the review?

File: applications/teletype/src/templates/teletype_account_zone_change.erl [hunk 5]
```diff
@@ -82,3 +88,12 @@ process_req(DataJObj) ->
         'ok' -> teletype_util:send_update(DataJObj, <<"completed">>);
         {'error', Reason} -> teletype_util:send_update(DataJObj, <<"failed">>, Reason)
     end.
+
+-spec zones_data(kz_json:object()) -> api_object().
```
Comment (jamesaimonetti) [inline applications/teletype/src/templates/teletype_account_zone_change.erl]: Spec is wrong. Return a JSON object in one clause and a proplist of one tuple in the other.

Comment (jamesaimonetti) [inline applications/teletype/src/templates/teletype_account_zone_change.erl]: I assume it needs to be `kz_proplist()`?

Comment (icehess) [inline applications/teletype/src/templates/teletype_account_zone_change.erl]: fixed

**Comment by icehess** [comment]
@jamesaimonetti @lazedo I don't know that you notice this or not, that I had break up `transaction` templates into two templates. Old transaction templates is still has the same id (`transaction`) with friendly_name set to "Transaction Success", and the second one is `transaction_failed` (friendly_name "Transaction Problem").

Also i changed the transaction notification message headers(removed "Transaction" and added some few necessary field instead). This is the change for both `transaction` and `topup`.

I don't know the impact of this changes, @k-anderson told me that it could be a problem.

Any idea?

**Comment by icehess** [comment]
Actually some of the macros in templates was changed (port_* and service_added for example).

Also There were some changes in braintree core library, like removing `refund_id` (I couldn't find any reference to it in Braintree docs), adding descriptor etc...

**Pull request merged**